stan/math/gpu/kernel_cl.hpp

```diff
@@ -0,0 +1,464 @@
+#ifndef STAN_MATH_GPU_CL_HPP
+#define STAN_MATH_GPU_CL_HPP
+#ifdef STAN_OPENCL
+#include <stan/math/gpu/opencl_context.hpp>
+#include <stan/math/gpu/constants.hpp>
+#include <CL/cl.hpp>
+#include <string>
+#include <algorithm>
+#include <map>
+#include <vector>
+
+namespace stan {
+namespace math {
+class kernel_cl;  // Declare here so kernel_cl_base can see the class
+namespace internal {
+class kernel_cl_base {
+  friend class stan::math::kernel_cl;
+
+ private:
+  std::string copy_matrix =
+#include <stan/math/gpu/kernels/copy_matrix.cl>  // NOLINT
+      ;                                          // NOLINT
+  std::string transpose_matrix =
+#include <stan/math/gpu/kernels/transpose_matrix.cl>  // NOLINT
+      ;                                               // NOLINT
+  std::string zeros_matrix =
+#include <stan/math/gpu/kernels/zeros_matrix.cl>  // NOLINT
+      ;                                           // NOLINT
+  std::string identity_matrix =
+#include <stan/math/gpu/kernels/identity_matrix.cl>  // NOLINT
+      ;                                              // NOLINT
+  std::string copy_triangular_matrix =
+#include <stan/math/gpu/kernels/copy_triangular_matrix.cl>  // NOLINT
+      ;                                                     // NOLINT
+  std::string copy_triangular_transposed_matrix =
+#include <stan/math/gpu/kernels/triangular_transpose.cl>  // NOLINT
+      ;                                                   // NOLINT
+  std::string copy_submatrix =
+#include <stan/math/gpu/kernels/sub_block.cl>  // NOLINT
+      ;                                        // NOLINT
+  std::string check_nan =
+#include <stan/math/gpu/kernels/check_nan.cl>  // NOLINT
+      ;                                        // NOLINT
+  std::string check_diagonal_zeros =
+#include <stan/math/gpu/kernels/check_diagonal_zeros.cl>  // NOLINT
+      ;                                                   // NOLINT
+  std::string check_symmetric =
+#include <stan/math/gpu/kernels/check_symmetric.cl>  // NOLINT
+      ;                                              // NOLINT
+  std::string subtract_symmetric =
+#include <stan/math/gpu/kernels/subtract_matrix.cl>  // NOLINT
+      ;                                              // NOLINT
+  std::string add_symmetric =
+#include <stan/math/gpu/kernels/add_matrix.cl>  // NOLINT
+      ;                                         // NOLINT
+  kernel_cl_base() {
+    std::string kernel_opts = "";
+    std::string kernel_source = helpers;
+    for (auto kernel_info : kernel_table) {
```

Comment:
I think the standard pattern I've seen is `auto&& kernel_info` so it's not copied into `kernel_info` on each iteration